File: base/src/AddressTable.cc
//
// Copyright (C) [2020] Futurewei Technologies, Inc.
//
// FORCE-RISCV is licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR
// FIT FOR A PARTICULAR PURPOSE.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include <AddressTable.h>
#include <Generator.h>
#include <RecoveryAddressGenerator.h>
#include <Register.h>
#include <Variable.h>
#include <AddressTableManager.h>
#include <Log.h>

#include <sstream>

using namespace std;

/*!
  \file AddressTable.cc
  \brief Code to implement address table for recovery and fast preamble addresses.
*/

namespace Force
{

  AddressTable::AddressTable(const EMemBankType bankType)
    : Object(), mpRecoveryAddrGenerator(nullptr), mpTableIndexRegister(nullptr), mBankType(bankType), mpTableSize(nullptr), mpGenerator(nullptr), mBaseAddress(0), mCurrentAddress(0), mCurrentTableSize(0)
  {
  }

  AddressTable::AddressTable(const AddressTable& rOther)
    : Object(rOther), mpRecoveryAddrGenerator(nullptr), mpTableIndexRegister(nullptr), mBankType(rOther.mBankType), mpTableSize(nullptr), mpGenerator(nullptr), mBaseAddress(0), mCurrentAddress(0), mCurrentTableSize(0)
  {
  }

  AddressTable::~AddressTable()
  {
    delete mpRecoveryAddrGenerator;
  }

  const std::string AddressTable::ToString() const
  {
    stringstream out_str;

    out_str << Type() << hex << " MemoryBank=" << EMemBankType_to_string(mBankType)
        << ",BaseAddress=0x"    << mBaseAddress
        << ",TableSize=0x"      << mCurrentTableSize
        << ",RegisterIndex=0x"  << mpTableIndexRegister->IndexValue()
        << ",RegisterValue=0x"  << mpTableIndexRegister->Value()
        << ",CurrentAddress=0x" << mCurrentAddress;

    return out_str.str();
  }

  void AddressTable::Setup(Generator* pGenerator)
  {
    mpGenerator = pGenerator;
    const VariableModerator* var_mod = mpGenerator->GetVariableModerator(EVariableType::Value);
    mpTableSize = var_mod->GetVariableSet()->FindVariable("Address table memory size");
  }

  void AddressTable::Initialize(cuint32 tableIndex, cuint64 baseAddress)
  {
    const RegisterFile* reg_file = mpGenerator->GetRegisterFile();
    mpTableIndexRegister = reg_file->RegisterLookupByIndex(tableIndex, ERegisterType::GPR, 64);
    mBaseAddress = GenerateNewTable(baseAddress);
    mCurrentAddress = mBaseAddress - 8;
  }

  void AddressTable::InitializeTableIndexRegister() const
  {
    if (mpTableIndexRegister->IsInitialized())
    {
      LOG(fail) << "{AddressTable::InitializeTableIndexRegister} recovery register already initialized" << endl;
      FAIL("reintialize-recovery-register");
    }

    mpGenerator->InitializeRegister(mpTableIndexRegister->Name(), "", mBaseAddress);
  }

  void AddressTable::GenerateAddress(cuint64 align, cuint64 size, cbool isInstr, const EMemAccessType memAccess, bool* pNewTable)
  {
    uint64 table_entry_addr = mpTableIndexRegister->Value();

    // If second to last entry in table, create a new table; then, write 1 followed by the address of the new table into
    // the existing table
    if (table_entry_addr == (mBaseAddress + mCurrentTableSize - 16))
    {
      uint64 new_table_addr = GenerateNewTable();
      mpGenerator->InitializeMemory(table_entry_addr, EMemBankTypeBaseType(mBankType), 8, 1, false, true);
      mpGenerator->InitializeMemory(table_entry_addr + 8, EMemBankTypeBaseType(mBankType), 8, new_table_addr, false, true);

      mBaseAddress = new_table_addr;
      mCurrentAddress = new_table_addr - 8;
      table_entry_addr = new_table_addr;

      if (pNewTable != nullptr) {
        (*pNewTable) = true;
      }
    }
    else if (pNewTable != nullptr) {
      (*pNewTable) = false;
    }

    // Generate a random VA and insert it into the table
    if (table_entry_addr == mCurrentAddress + 8)
    {
      uint64 table_entry_val = mpRecoveryAddrGenerator->GenerateAddress(align, size, isInstr, memAccess);
      mpGenerator->InitializeMemory(table_entry_addr, EMemBankTypeBaseType(mBankType), 8, table_entry_val, false, true);
      mCurrentAddress = table_entry_addr;

      LOG(notice) << "{AddressTable::GenerateAddress} generated addr=0x" << hex << table_entry_val <<
        " at addr=0x" << table_entry_addr << " memory bank=" << EMemBankType_to_string(mBankType) << endl;
    }
  }

  void AddressTable::SetRecoveryAddressGenerator(const RecoveryAddressGenerator* pRecoveryAddrGenerator)
  {
    if (mpRecoveryAddrGenerator != nullptr) {
      LOG(fail) << "{AddressTable::SetRecoveryAddressGenerator} recovery address generator already set" << endl;
      FAIL("reset-recovery-addr-generator");
    }

    mpRecoveryAddrGenerator = pRecoveryAddrGenerator;
  }

  uint64 AddressTable::GenerateNewTable(cuint64 baseAddress)
  {
    mCurrentTableSize = mpTableSize->Value();
    return mpGenerator->GetAddressTableManager()->GenerateNewTable(baseAddress, mCurrentTableSize, mBankType);
  }

}

File: py/riscv/AddressTable.py
#
# Copyright (C) [2020] Futurewei Technologies, Inc.
#
# FORCE-RISCV is licensed under the Apache License, Version 2.0
#  (the "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES
# OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
# NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from base.Sequence import Sequence
import inspect
import sys
import os

## AddressTableRISCV
# Provide available address for exceptions handlers.
class AddressTableRISCV(Sequence):

    def __init__(self, gen_thread):
        super().__init__(gen_thread)
        self.table_index = None
        self._mAppRegSize = 64

    def generate(self, **kwargs):
        self.table_index = kwargs.get('table_index', None)
        if self.table_index is None:
            self.table_index = self.getRandomGPR(exclude="0,1,2")
        self.reserveRegisterByIndex(64,self.table_index,"GPR","ReadWrite")
        self._mAppRegSize = self.getGlobalState('AppRegisterWidth')
        
        return self.table_index

    def tableIndex(self):
        return self.table_index

    def getAddress(self, reg_index, scratch_index):
        #Load the value pointed to by the reg table_index into reg reg_index and then increment the value in the reg table_index by 8
        self.genLoadRegFromDataBlock(reg_index, self.table_index, True)
        
        #get scratch GPR and load it with the number 1 
        self.genInstruction('ADDI##RISCV', {'rd': scratch_index, 'rs1': 0, 'simm12': 1})

        #Compare the value that had been loaded into reg reg_index to one. One means the table entry after this one is actually the address of another table.
        pc = self.getPEstate("PC")
        self.genInstruction('BNE##RISCV', {'rs1': scratch_index, 'rs2': reg_index, 'simm12': 8, 'NoBnt': 1, 'NoRestriction': 1})
        self.setPEstate("PC", pc + 4)

        #Load the value pointed to by the reg table_index into reg table_index.
        self.genLoadRegFromDataBlock(self.table_index, self.table_index)

        #Load the value pointed to by the reg table_index into reg reg_index and then increment the value in the reg table_index by 8
        self.genLoadRegFromDataBlock(reg_index, self.table_index, True)

    def genLoadRegFromDataBlock(self, aDestReg, aAddrReg, aIncrementAddr = False):
        if self._mAppRegSize == 32:
            self.genInstruction('LW##RISCV', {'rd': aDestReg, 'rs1': aAddrReg, 'simm12': 0, 'NoRestriction': 1})
            if aIncrementAddr:
                self.genInstruction('ADDI##RISCV', {'rd': aAddrReg, 'rs1': aAddrReg, 'simm12': 4})
        else:
            self.genInstruction('LD##RISCV', {'rd': aDestReg, 'rs1': aAddrReg, 'simm12': 0, 'NoRestriction': 1})
            if aIncrementAddr:
                self.genInstruction('ADDI##RISCV', {'rd': aAddrReg, 'rs1': aAddrReg, 'simm12': 8})

## AddressTableManagerRISCV class
class AddressTableManagerRISCV(Sequence):

    def __init__(self, gen_thread):
        super().__init__(gen_thread)
        self.table_index = None        # address table index
        self.address_table = None      # address table

    def generate(self):
        self.address_table = AddressTableRISCV(self.genThread)
        self.table_index = self.address_table.generate(table_index=self.table_index)
        fast_mode = self.genThread.fastMode 
        addr_table_info_set = {}
        addr_table_info_set["table_index"] = self.table_index;
        addr_table_info_set["fast_mode"] = fast_mode 
        self.genSequence("InitializeAddrTables", addr_table_info_set)

    # only copy table index.
    def createShallowCopy(self, gen_thread):
        address_table_manager_copy = AddressTableManagerRISCV(gen_thread)
        address_table_manager_copy.table_index = self.table_index
        return address_table_manager_copy

    def addressTable(self):
        return self.address_table
